### PR TITLE
add ALLOW_PLAYER_CHANGE_WORLD

### PR DESCRIPTION
**Pull request by fishshi**
ALLOW_PLAYER_CHANGE_WORLD: Called when the player's dimension changes to check if the player is allowed to cross dimensions.

solve issue:
https://github.com/FabricMC/fabric/issues/3294#issue-1877370481

### CODE REVIEW

File: fabric-entity-events-v1/src/testmod/java/net/fabricmc/fabric/test/entity/event/EntityEventTests.java [hunk 1]
```diff
@@ -64,6 +64,16 @@ public void onInitialize() {
 			LOGGER.info("Entity {} Killed: {}", entity, killed);
 		});
 
+		ServerEntityWorldChangeEvents.ALLOW_PLAYER_CHANGE_WORLD.register((player, origin, destination) -> {
+			if (player.getStackInHand(Hand.MAIN_HAND).getItem() == Items.END_ROD) {
```
Comment (haykam821) [inline fabric-entity-events-v1/src/testmod/java/net/fabricmc/fabric/test/entity/event/EntityEventTests.java]: Nitpick:

```suggestion
			if (player.getMainHandStack().isOf(Items.END_ROD)) {
```

File: fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/api/entity/event/v1/ServerEntityWorldChangeEvents.java [hunk 2]
```diff
@@ -45,6 +45,19 @@ public final class ServerEntityWorldChangeEvents {
 		}
 	});
 
+	/**
+	 * Called before a player is being moved to a different world.
+	 */
+	public static final Event<AllowPlayerChange> ALLOW_PLAYER_CHANGE_WORLD = EventFactory.createArrayBacked(AllowPlayerChange.class, callbacks -> (player, origin, destination) -> {
+		for (AllowPlayerChange callback : callbacks) {
+			if (!callback.allowChangeWorld(player, origin, destination)){
```
Comment (haykam821) [inline fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/api/entity/event/v1/ServerEntityWorldChangeEvents.java]: Checkstyle will fail:

```suggestion
			if (!callback.allowChangeWorld(player, origin, destination)) {
```

**Review by haykam821 — CHANGES_REQUESTED** [review]

File: fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/api/entity/event/v1/ServerEntityWorldChangeEvents.java (hunk 2 above)
Comment (fishshi) [reply fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/api/entity/event/v1/ServerEntityWorldChangeEvents.java]: thank you!

**Comment by fishshi** [comment]
Sorry, but I accidentally merged the base branch into this branch. Will this have any impact, and how can I recover from it?

File: fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/api/entity/event/v1/ServerEntityWorldChangeEvents.java [hunk 3]
```diff
@@ -75,6 +88,19 @@ public interface AfterEntityChange {
 		void afterChangeWorld(Entity originalEntity, Entity newEntity, ServerWorld origin, ServerWorld destination);
 	}
 
+	@FunctionalInterface
+	public interface AllowPlayerChange {
```
Comment (modmuss50) [inline fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/api/entity/event/v1/ServerEntityWorldChangeEvents.java:92]: Should this be limited to just players, other entities are allowed to use portals.

I think its also worth looking at 1.21.2 as I know they changed a lot of this teleportation code.

Comment (fishshi) [reply fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/api/entity/event/v1/ServerEntityWorldChangeEvents.java:92]: Yes, this event is currently only used to restrict players when crossing dimensions; other entities are not affected. I believe there are very few scenarios where restricting entities from crossing dimensions would be necessary.

Comment (fishshi) [reply fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/api/entity/event/v1/ServerEntityWorldChangeEvents.java:92]: Additionally, I will read through 1.21.2 as soon as possible, and I may be able to provide feedback by tomorrow.

Comment (fishshi) [reply fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/api/entity/event/v1/ServerEntityWorldChangeEvents.java:92]: I have reviewed the relevant source code for 24w39a, and there are not many changes regarding the dimension-crossing code. Through testing, I can assure you that this event works correctly in the latest version as well.

**Comment by TelepathicGrunt** [comment]
Big concern with this PR. Say you block a teleport from a mod but that mod still sends the packets relating to teleporting. Won’t that cause issues and weird states in the game? For example, if you block the teleportTo in Bumblezone, my other packets will still be sent. https://github.com/TelepathicGrunt/Bumblezone/blob/2dc4c2034f1d9caf52cbe4c798754721b4895039/common/src/main/java/com/telepathicgrunt/the_bumblezone/entities/teleportation/BzWorldSavedData.java#L314

This is probably gonna cause major problems for the client and possibly other people as well

**Comment by modmuss50** [comment]
Hi, thanks for the PR unfortunately I think doing this well requires quite a lot of work, teleportation is really complex. Some things that you will likely need to take into account:

- None player entities
- The context of how the entity is being teleported. E.g was it a command, portal, death, end credits or something another mod has initiated.
- There should be 3 events, allow, on and canceled.
- What state will the client and the entity being teleported be left in when the teleportation is blocked.

The first thing to do is to likely define a clear set of requirements/use cases, this likely requires looking at what existing mods do.

**Comment by fishshi** [comment]
I'm really sorry, but I've decided to give up on this. Anyone else is welcome to continue trying.

**Pull request closed without merging**